File: python/src/lichen/senml/ipso.py
# SPDX-License-Identifier: GPL-3.0-or-later
# SPDX-FileCopyrightText: The contributors to the LICHEN project
"""IPSO Smart Object names for SenML records.

IPSO/LwM2M identifies a resource with the numeric path
``object-id/object-instance-id/resource-id``.  SenML names are relative, so
LICHEN carries the same path without a leading slash.  The two-component
``object-id/object-instance-id`` form is also supported for composite profiles
where the resource is implied by the profile.
"""

from __future__ import annotations

import math
from collections.abc import Mapping
from dataclasses import dataclass
from decimal import Decimal
from enum import IntEnum
from types import MappingProxyType

from lichen.senml.codec import SenmlRecord

_MAX_ID = 0xFFFF


class IpsoObjectId(IntEnum):
    """IPSO Smart Object identifiers used by the LICHEN sensor profiles."""

    TEMPERATURE = 3303
    HUMIDITY = 3304
    ACCELEROMETER = 3313
    BAROMETER = 3315
    PRESSURE = 3323
    GYROMETER = 3334
    LOCATION = 3336


class IpsoResourceId(IntEnum):
    """Reusable IPSO resource identifiers used by supported objects."""

    TIMESTAMP = 5518
    SENSOR_VALUE = 5700
    SENSOR_UNITS = 5701
    X_VALUE = 5702
    Y_VALUE = 5703
    Z_VALUE = 5704
    COMPASS_DIRECTION = 5705
    NUMERIC_LATITUDE = 6051
    NUMERIC_LONGITUDE = 6052
    NUMERIC_UNCERTAINTY = 6053


@dataclass(frozen=True, slots=True)
class IpsoObjectDefinition:
    """Known object metadata needed to create its primary sensor record."""

    object_id: IpsoObjectId
    name: str
    value_resource_id: IpsoResourceId
    default_unit: str


_OBJECT_DEFINITIONS: dict[IpsoObjectId, IpsoObjectDefinition] = {
    IpsoObjectId.TEMPERATURE: IpsoObjectDefinition(
        IpsoObjectId.TEMPERATURE, "Temperature", IpsoResourceId.SENSOR_VALUE, "Cel"
    ),
    IpsoObjectId.HUMIDITY: IpsoObjectDefinition(
        IpsoObjectId.HUMIDITY, "Humidity", IpsoResourceId.SENSOR_VALUE, "%RH"
    ),
    IpsoObjectId.ACCELEROMETER: IpsoObjectDefinition(
        IpsoObjectId.ACCELEROMETER, "Accelerometer", IpsoResourceId.X_VALUE, "m/s2"
    ),
    IpsoObjectId.BAROMETER: IpsoObjectDefinition(
        IpsoObjectId.BAROMETER, "Barometer", IpsoResourceId.SENSOR_VALUE, "Pa"
    ),
    IpsoObjectId.PRESSURE: IpsoObjectDefinition(
        IpsoObjectId.PRESSURE, "Pressure", IpsoResourceId.SENSOR_VALUE, "Pa"
    ),
    IpsoObjectId.GYROMETER: IpsoObjectDefinition(
        IpsoObjectId.GYROMETER, "Gyrometer", IpsoResourceId.X_VALUE, "rad/s"
    ),
    IpsoObjectId.LOCATION: IpsoObjectDefinition(
        IpsoObjectId.LOCATION, "Location", IpsoResourceId.NUMERIC_LATITUDE, "lat"
    ),
}

IPSO_OBJECTS: Mapping[IpsoObjectId, IpsoObjectDefinition] = MappingProxyType(
    _OBJECT_DEFINITIONS
)

_RESOURCE_UNITS: Mapping[tuple[IpsoObjectId, IpsoResourceId], str] = MappingProxyType(
    {
        (definition.object_id, definition.value_resource_id): definition.default_unit
        for definition in _OBJECT_DEFINITIONS.values()
    }
    | {
        (IpsoObjectId.ACCELEROMETER, IpsoResourceId.Y_VALUE): "m/s2",
        (IpsoObjectId.ACCELEROMETER, IpsoResourceId.Z_VALUE): "m/s2",
        (IpsoObjectId.GYROMETER, IpsoResourceId.Y_VALUE): "rad/s",
        (IpsoObjectId.GYROMETER, IpsoResourceId.Z_VALUE): "rad/s",
        (IpsoObjectId.LOCATION, IpsoResourceId.NUMERIC_LONGITUDE): "lon",
        (IpsoObjectId.LOCATION, IpsoResourceId.NUMERIC_UNCERTAINTY): "m",
        (IpsoObjectId.LOCATION, IpsoResourceId.COMPASS_DIRECTION): "deg",
    }
)


def _validate_id(label: str, value: int) -> int:
    if isinstance(value, bool) or not isinstance(value, int):
        raise TypeError(f"{label} must be an integer")
    if not (0 <= value <= _MAX_ID):
        raise ValueError(f"{label} {value} out of range [0, 65535]")
    return int(value)


def _parse_component(label: str, component: str) -> int:
    if not component or not component.isascii() or not component.isdecimal():
        raise ValueError(f"{label} must be an unsigned decimal integer")
    value = int(component)
    _validate_id(label, value)
    if str(value) != component:
        raise ValueError(f"{label} must use canonical decimal encoding")
    return value


@dataclass(frozen=True, slots=True)
class IpsoPath:
    """A canonical IPSO object/instance[/resource] relative path."""

    object_id: int
    instance_id: int = 0
    resource_id: int | None = None

    def __post_init__(self) -> None:
        _validate_id("object_id", self.object_id)
        _validate_id("instance_id", self.instance_id)
        if self.resource_id is not None:
            _validate_id("resource_id", self.resource_id)

    def __str__(self) -> str:
        path = f"{self.object_id}/{self.instance_id}"
        if self.resource_id is not None:
            path += f"/{self.resource_id}"
        return path

    @classmethod
    def parse(cls, name: str) -> IpsoPath:
        """Parse a canonical relative IPSO name.

        Leading slashes, signs, whitespace, leading zeroes, missing components,
        and identifiers outside the LwM2M 16-bit identifier space are rejected.
        """
        if not isinstance(name, str):
            raise TypeError("IPSO path must be a string")
        components = name.split("/")
        if len(components) not in (2, 3):
            raise ValueError("IPSO path must have 2 or 3 components")
        object_id = _parse_component("object_id", components[0])
        instance_id = _parse_component("instance_id", components[1])
        resource_id = (
            _parse_component("resource_id", components[2]) if len(components) == 3 else None
        )
        return cls(object_id, instance_id, resource_id)


def object_definition(object_id: int) -> IpsoObjectDefinition | None:
    """Return metadata for a supported object ID, or ``None`` if unknown."""
    _validate_id("object_id", object_id)
    try:
        known_id = IpsoObjectId(object_id)
    except ValueError:
        return None
    return IPSO_OBJECTS[known_id]


def default_unit(object_id: int, resource_id: int) -> str | None:
    """Return the LICHEN SenML unit for a known object/resource pair."""
    _validate_id("object_id", object_id)
    _validate_id("resource_id", resource_id)
    try:
        key = (IpsoObjectId(object_id), IpsoResourceId(resource_id))
    except ValueError:
        return None
    return _RESOURCE_UNITS.get(key)


def sensor_record(
    object_id: int,
    value: int | float | Decimal,
    *,
    instance_id: int = 0,
    resource_id: int | None = None,
    unit: str | None = None,
) -> SenmlRecord:
    """Create a numeric SenML record named with an IPSO resource path.

    If ``resource_id`` is omitted, known objects use their primary value
    resource and unknown objects use the reusable Sensor Value resource 5700.
    Units are inferred only for known object/resource pairs; an explicit
    ``unit`` always wins.
    """
    if isinstance(value, bool) or not isinstance(value, (int, float, Decimal)):
        raise TypeError("value must be a number")
    try:
        finite = math.isfinite(value)
    except OverflowError:
        finite = False
    if not finite:
        raise ValueError("value must be finite")

    definition = object_definition(object_id)
    if resource_id is None:
        resource_id = (
            int(definition.value_resource_id)
            if definition is not None
            else int(IpsoResourceId.SENSOR_VALUE)
        )
    path = IpsoPath(object_id, instance_id, resource_id)
    record_unit = unit if unit is not None else default_unit(object_id, resource_id)
    return SenmlRecord(n=str(path), u=record_unit, v=value)

File: rust/lichen-senml/src/ipso.rs
//! IPSO Smart Object names for SenML records.
//!
//! LwM2M resource paths are `object/instance/resource`. SenML names use the
//! same path without the CoAP URI's leading slash. The compact two-component
//! `object/instance` form is available when a composite profile implies the
//! resource.

use core::fmt;

use crate::Record;

/// IPSO Smart Object identifiers used by the LICHEN sensor profiles.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u16)]
pub enum ObjectId {
    Temperature = 3303,
    Humidity = 3304,
    Accelerometer = 3313,
    Barometer = 3315,
    Pressure = 3323,
    Gyrometer = 3334,
    Location = 3336,
}

/// Reusable IPSO resources used by supported objects.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u16)]
pub enum ResourceId {
    Timestamp = 5518,
    SensorValue = 5700,
    SensorUnits = 5701,
    XValue = 5702,
    YValue = 5703,
    ZValue = 5704,
    CompassDirection = 5705,
    NumericLatitude = 6051,
    NumericLongitude = 6052,
    NumericUncertainty = 6053,
}

/// Metadata needed to create the primary SenML record for a known object.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct ObjectDefinition {
    pub id: ObjectId,
    pub name: &'static str,
    pub value_resource: ResourceId,
    pub default_unit: &'static str,
}

/// Objects for which LICHEN defines SenML units and primary resources.
pub const OBJECTS: &[ObjectDefinition] = &[
    ObjectDefinition {
        id: ObjectId::Temperature,
        name: "Temperature",
        value_resource: ResourceId::SensorValue,
        default_unit: "Cel",
    },
    ObjectDefinition {
        id: ObjectId::Humidity,
        name: "Humidity",
        value_resource: ResourceId::SensorValue,
        default_unit: "%RH",
    },
    ObjectDefinition {
        id: ObjectId::Accelerometer,
        name: "Accelerometer",
        value_resource: ResourceId::XValue,
        default_unit: "m/s2",
    },
    ObjectDefinition {
        id: ObjectId::Barometer,
        name: "Barometer",
        value_resource: ResourceId::SensorValue,
        default_unit: "Pa",
    },
    ObjectDefinition {
        id: ObjectId::Pressure,
        name: "Pressure",
        value_resource: ResourceId::SensorValue,
        default_unit: "Pa",
    },
    ObjectDefinition {
        id: ObjectId::Gyrometer,
        name: "Gyrometer",
        value_resource: ResourceId::XValue,
        default_unit: "rad/s",
    },
    ObjectDefinition {
        id: ObjectId::Location,
        name: "Location",
        value_resource: ResourceId::NumericLatitude,
        default_unit: "lat",
    },
];

/// Return metadata for a supported object ID.
pub fn object_definition(object_id: u16) -> Option<&'static ObjectDefinition> {
    OBJECTS
        .iter()
        .find(|definition| definition.id as u16 == object_id)
}

/// Return the LICHEN SenML unit for a known object/resource pair.
pub fn default_unit(object_id: u16, resource_id: u16) -> Option<&'static str> {
    match (object_id, resource_id) {
        (3303, 5700) => Some("Cel"),
        (3304, 5700) => Some("%RH"),
        (3313, 5702..=5704) => Some("m/s2"),
        (3315 | 3323, 5700) => Some("Pa"),
        (3334, 5702..=5704) => Some("rad/s"),
        (3336, 6051) => Some("lat"),
        (3336, 6052) => Some("lon"),
        (3336, 6053) => Some("m"),
        (3336, 5705) => Some("deg"),
        _ => None,
    }
}

/// A canonical IPSO object/instance[/resource] relative path.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Path {
    pub object_id: u16,
    pub instance_id: u16,
    pub resource_id: Option<u16>,
}

impl Path {
    /// Create a compact object/instance name.
    pub const fn object_instance(object_id: u16, instance_id: u16) -> Self {
        Self {
            object_id,
            instance_id,
            resource_id: None,
        }
    }

    /// Create a full object/instance/resource name.
    pub const fn resource(object_id: u16, instance_id: u16, resource_id: u16) -> Self {
        Self {
            object_id,
            instance_id,
            resource_id: Some(resource_id),
        }
    }

    /// Create the primary value path for a known object.
    pub fn primary(object_id: u16, instance_id: u16) -> Self {
        let resource_id = object_definition(object_id)
            .map(|definition| definition.value_resource as u16)
            .unwrap_or(ResourceId::SensorValue as u16);
        Self::resource(object_id, instance_id, resource_id)
    }

    /// Parse a canonical relative IPSO path.
    pub fn parse(name: &str) -> Result<Self, PathError> {
        let mut components = name.split('/');
        let object = components.next().ok_or(PathError::InvalidShape)?;
        let instance = components.next().ok_or(PathError::InvalidShape)?;
        let resource = components.next();
        if components.next().is_some() {
            return Err(PathError::InvalidShape);
        }
        if resource.is_none() && (object.is_empty() || instance.is_empty()) {
            return Err(PathError::InvalidComponent);
        }
        Ok(Self {
            object_id: parse_component(object)?,
            instance_id: parse_component(instance)?,
            resource_id: resource.map(parse_component).transpose()?,
        })
    }

    /// Number of ASCII bytes required by the canonical name.
    pub const fn encoded_len(&self) -> usize {
        let mut len = decimal_len(self.object_id) + 1 + decimal_len(self.instance_id);
        if let Some(resource_id) = self.resource_id {
            len += 1 + decimal_len(resource_id);
        }
        len
    }

    /// Write the canonical name into caller-provided storage.
    pub fn write_name<'a>(&self, out: &'a mut [u8]) -> Result<&'a str, NameError> {
        let needed = self.encoded_len();
        if out.len() < needed {
            return Err(NameError::BufferTooSmall {
                needed,
                available: out.len(),
            });
        }

        let mut position = 0;
        position += write_decimal(self.object_id, &mut out[position..]);
        out[position] = b'/';
        position += 1;
        position += write_decimal(self.instance_id, &mut out[position..]);
        if let Some(resource_id) = self.resource_id {
            out[position] = b'/';
            position += 1;
            position += write_decimal(resource_id, &mut out[position..]);
        }

        core::str::from_utf8(&out[..position]).map_err(|_| NameError::InvalidUtf8)
    }

    /// Build a numeric SenML record using caller-owned storage for the name.
    pub fn record<'a>(
        &self,
        name_storage: &'a mut [u8],
        value: f64,
        unit: Option<&'a str>,
    ) -> Result<Record<'a>, NameError> {
        let name = self.write_name(name_storage)?;
        Ok(Record {
            name: Some(name),
            value: Some(value),
            unit,
            ..Record::empty()
        })
    }

    /// Unit inferred for this exact known object/resource pair.
    pub fn default_unit(&self) -> Option<&'static str> {
        self.resource_id
            .and_then(|resource_id| default_unit(self.object_id, resource_id))
    }
}

impl fmt::Display for Path {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(formatter, "{}/{}", self.object_id, self.instance_id)?;
        if let Some(resource_id) = self.resource_id {
            write!(formatter, "/{resource_id}")?;
        }
        Ok(())
    }
}

/// IPSO path parsing failure.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PathError {
    InvalidShape,
    InvalidComponent,
    NonCanonical,
    OutOfRange,
}

impl fmt::Display for PathError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::InvalidShape => formatter.write_str("IPSO path must have 2 or 3 components"),
            Self::InvalidComponent => {
                formatter.write_str("IPSO path components must be unsigned decimal integers")
            }
            Self::NonCanonical => {
                formatter.write_str("IPSO path components must use canonical decimal encoding")
            }
            Self::OutOfRange => formatter.write_str("IPSO path component exceeds 65535"),
        }
    }
}

impl core::error::Error for PathError {}

/// IPSO name formatting failure.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum NameError {
    BufferTooSmall { needed: usize, available: usize },
    InvalidUtf8,
}

impl fmt::Display for NameError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::BufferTooSmall { needed, available } => write!(
                formatter,
                "IPSO name buffer too small: need {needed} bytes, have {available}"
            ),
            Self::InvalidUtf8 => formatter.write_str("IPSO name was not valid UTF-8"),
        }
    }
}

impl core::error::Error for NameError {}

fn parse_component(component: &str) -> Result<u16, PathError> {
    if component.is_empty() || !component.bytes().all(|byte| byte.is_ascii_digit()) {
        return Err(PathError::InvalidComponent);
    }
    if component.len() > 1 && component.starts_with('0') {
        return Err(PathError::NonCanonical);
    }
    let mut value = 0_u32;
    for byte in component.bytes() {
        value = value
            .checked_mul(10)
            .and_then(|current| current.checked_add(u32::from(byte - b'0')))
            .ok_or(PathError::OutOfRange)?;
        if value > u32::from(u16::MAX) {
            return Err(PathError::OutOfRange);
        }
    }
    Ok(value as u16)
}

const fn decimal_len(value: u16) -> usize {
    match value {
        0..=9 => 1,
        10..=99 => 2,
        100..=999 => 3,
        1000..=9999 => 4,
        _ => 5,
    }
}

fn write_decimal(value: u16, out: &mut [u8]) -> usize {
    let len = decimal_len(value);
    let mut remaining = value;
    let mut position = len;
    while position > 0 {
        position -= 1;
        out[position] = b'0' + (remaining % 10) as u8;
        remaining /= 10;
    }
    len
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn maximum_path_fits_seventeen_bytes() {
        let path = Path::resource(u16::MAX, u16::MAX, u16::MAX);
        let mut name = [0_u8; 17];
        assert_eq!(path.write_name(&mut name), Ok("65535/65535/65535"));
    }

    #[test]
    fn record_borrows_formatted_name() {
        let path = Path::primary(ObjectId::Temperature as u16, 0);
        let mut name = [0_u8; 17];
        let record = path.record(&mut name, 23.5, path.default_unit()).unwrap();
        assert_eq!(record.name, Some("3303/0/5700"));
        assert_eq!(record.unit, Some("Cel"));
        assert_eq!(record.value, Some(23.5));
    }
}
